lib/model/folder_sendrecv.go

```diff
@@ -1334,16 +1294,93 @@ func (f *sendReceiveFolder) copierRoutine(in <-chan copyBlocksState, pullChan ch
 				state.copyDone(block)
 			}
 		}
-		if file != nil {
-			// os.File used to return invalid argument if nil.
-			// fs.File panics as it's an interface.
-			file.Close()
-		}
 
 		out <- state.sharedPullerState
 	}
 }
 
+// copyByWeakHash copies all blocks that can be found by their weak hash.
+// It returns an updated state with these blocks removed.
+func (f *sendReceiveFolder) copyByWeakHash(state copyBlocksState, w io.WriterAt) (copyBlocksState, error) {
+	file, err := f.fs.Open(state.file.Name)
+	if err != nil {
+		return state, err
+	}
+	defer file.Close()
+
+	buf := protocol.BufferPool.Get(state.file.BlockSize())
+	defer protocol.BufferPool.Put(buf)
+
+	finder := weakhash.NewFinder(file, buf)
+
+	byWeakhash := make(map[uint32][]int) // Block indices by weak hash.
+	for i, block := range state.blocks {
+		if int(block.Size) != len(buf) {
+			// Skip the last block if it isn't block.Size bytes.
```

Comment:
```suggestion
			// Skip the last block if it isn'tstate.file.BlockSize() bytes.
```

Reply:
I guess this is a conscious decision, but it feels it wouldn't add much complexity to support non BlockSize sized blocks?

Reply:
That would require maintaining two Adler-32 rolling hashes (so two sliding windows). It can be done, but the gain is always at most one fewer block to pull.